### ISSUE TITLE
bit-perfect streaming

### ISSUE THREAD

**Issue by papavlos** [ugjka/blast#18]
Hi @ugjka 

Thanks for the **blast**! First WORKING solution to stream audio from PipeWire to a remote DLNA renderer! Awesome!!!

I have the following enhancement request:

I have configured blast to stream losslessly by option `-usewav` and proper `-rate` and `-bits` values. That means that blast is sending the stream without any compression, but always resampling everything to upfront fixed samplerate/depth, unless you play the stream of the same samplerate/depth...

I am wondering if it would be possible to create a new option/options for blast, so then  blast is simply only re-sending the stream of the same bitsdepth and samplerate as it receives at the moment, without any resampling. 
For instance a new option `-bitperfect`, or extending existing options to accept "auto" values, like `-rate auto` and `-bits auto`.

I would appreciate your feedback.

**Comment by ugjka** [comment]
I have trouble believing that the DLNA receivers support changing format specifics arbitrary midstream.

What i can do with parec is this though
```
 --fix-format
              If passed, the sample format of the stream is changed to the  native  format  of  the  sink  the
              stream is connected to.

--fix-rate
              If  passed, the sampling rate of the stream is changed to the native rate of the sink the stream
              is connected to.

--fix-channels
              If passed, the number of channels and the channel map of the stream is  changed  to  the  native
              number of channels and the native channel map of the sink the stream is connected to.
```

So I could make it pick whatever your system is configured to use by default be it pipewire or pulseaudio

**Comment by papavlos** [comment]
I think that most of the hi-fi network streaming devices have the possibility of changing bitrate of their DLNA renderers on-the-fly, following the incoming stream. You may have a playlist composed of flac files with different sampling frequencies and the renderer captures and sends it bit-perfectly to the DAC attached, what you can observe on the DAC - changing frequencies and bit depths.
A sample of a client sending it accordingly is foobar2000 with foo_out_upnp component. This is somehow an equivalent - to blast - solution for Windows. And this combination of fb2k/foo_out_upnp works also under WINE in Linux, completely bypassing system (Windows or Linux) sound stack.
There are also players in Linux (Strawberry for example) that can bypass system resampling and send direct audio streams to DAC attached via USB, but only when operating directly through ALSA.

Coming to your proposal with `--fix-***` options: will it work when the sink attached changes its rate and bit on-the-fly, following the multimedia playing application?
I am not a "go" programmer, but when looking at "main.go" - don't you fix all parameter values at program start? By reading commandline arguments and setting given/default values?
If "blast" may somehow change its parameters following the stream to the sink - that would by my desired functionality. :)

**Comment by ugjka** [comment]
I'll experiment when I have some free time, will see where this goes

**Comment by ugjka** [comment]
> [ugjka@ugjka blast]$ sgpt "does wav format allow chaning samplerate and bitdepth on the fly"
> No, the WAV format does not support changing the sample rate and bit depth on the fly within a single file. WAV
> files have a header that specifies the sample rate, bit depth, and other parameters for the entire file. If you
> need to change these parameters, you would typically need to create a new WAV file with the desired settings.

Pretty sure your mentioned foobar2000 works on file level not in streaming mode

**Comment by papavlos** [comment]
And the sgpt doesn't lie. Of course, every single lossless PCM-based file has the same sample rate and bit depth for all samples in the given file, set up upfront.
Did I say earlier anything different?

I am talking about a playlist consisted of files having mixture of these parameters.
I would like blast to switch the bit depth and sample rate automatically, depending on  the track/file in the playlist. Between the tracks. After the previous track stops and before the new track starts. Whatever a player (like fb2k or Strawberry or VLC) sends to the sink the blast is attached to. So the blast doesn't resample anything. Just observes the parameters of the next incoming stream and a accommodates accordingly, before starting the network transmission to the DLNA renderer. Not during the stream, but between tracks - which means between streams.

Currently I would have to stop the blast daemon service everytime the audio track changes its parameters and start it with new ones. I would like blast to change sample rate and bit depth on-the-fly, without service restarting.

**Comment by ugjka** [comment]
I don't think this is really possible with the current architecture.

**Comment by papavlos** [comment]
OK. Thanks!
Maybe one day it comes to blast redesign and refactoring, so the please keep it in mind.

I have another idea for blast to make it capable for starting as a systemd service with the system start, but this is another story. Not for this thread. And not so important now.

Take care!
papavlos

**Comment by ugjka** [comment]
I'll keep this in mind, closing...
